src/crypto/generichash/mod.rs

```diff
@@ -42,74 +42,80 @@ impl State {
     /// `key` is an optional parameter, which when given,
     /// a custom key can be used for the computation of the hash.
     /// The size of the key must be in the interval [`KEY_MIN`, `KEY_MAX`].
-    pub fn new(out_len: usize, key: Option<&[u8]>) -> Option<State> {
+    pub fn new(out_len: usize, key: Option<&[u8]>) -> Result<State, ()> {
         if out_len < DIGEST_MIN || out_len > DIGEST_MAX {
-            return None;
+            return Err(());
         }
 
         if let Some(key) = key {
             let len = key.len();
             if len < KEY_MIN || len > KEY_MAX {
-                return None;
+                return Err(());
             }
         }
 
         let mut state: crypto_generichash_state;
 
         let result = unsafe {
             state = mem::uninitialized();
-            let state_ptr = &mut state as *mut _;
             if let Some(key) = key {
                 crypto_generichash_init(
-                    state_ptr,
+                    &mut state,
                     key.as_ptr(),
                     key.len(),
                     out_len,
                 )
             } else {
-                crypto_generichash_init(state_ptr, ptr::null(), 0, out_len)
+                crypto_generichash_init(&mut state, ptr::null(), 0, out_len)
             }
         };
 
         if result == 0 {
-            Some(State {
+            Ok(State {
                 out_len: out_len,
                 state,
             })
         } else {
-            None
+            Err(())
         }
     }
 
     /// `update` updates the `State` with `data`. `update` can be called multiple times in order
     /// to compute the hash from sequential chunks of the message.
-    pub fn update(&mut self, data: &[u8]) {
-        unsafe {
-            let state_ptr = &mut self.state as *mut _;
+    pub fn update(&mut self, data: &[u8]) -> Result<(), ()> {
+        if unsafe {
```

Comment:
Personally, I find it a little awkward how you wrote a predicate that spans 7 lines instead of going with `let rc = unsafe { ...` and then putting `if rc == 0 { ....` below, but it's still definitely followable and I don't think it needs changing unless you want.

Reply:
Should be fixed.